homeassistant/components/heos/__init__.py

```diff
@@ -25,28 +26,59 @@
 
 
 async def async_setup(hass: HomeAssistantType, config: ConfigType):
-    """Set up the HEOS component."""
-    from aioheos import AioHeosController
-
+    """Create config entry from config if not already created."""
     host = config[DOMAIN][CONF_HOST]
-    controller = AioHeosController(hass.loop, host)
+    entries = hass.config_entries.async_entries(DOMAIN)
+    if not entries:
+        # Create new entry based on config
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={'source': 'import'},
+                data={CONF_HOST: host}))
+    else:
+        # Check if host needs to be updated
+        entry = entries[0]
+        if entry.data[CONF_HOST] != host:
+            entry.data[CONF_HOST] = host
+            entry.title = format_title(host)
+            hass.config_entries.async_update_entry(entry)
 
-    try:
-        await asyncio.wait_for(controller.connect(), timeout=5.0)
-    except asyncio.TimeoutError:
-        _LOGGER.error('Timeout during setup.')
-        return False
+    return True
 
-    async def controller_close(event):
-        """Close connection when HASS shutsdown."""
-        await controller.close()
 
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, controller_close)
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Initialize config entry which represents the HEOS controller."""
+    from pyheos import Heos
+    # Setting all_progress_events=False ensures that we only receive a
+    # media position update upon start of playback or when media changes
+    controller = Heos(entry.data[CONF_HOST], all_progress_events=False)
 
-    hass.data.setdefault(DOMAIN, {})
-    hass.data[DOMAIN][MEDIA_PLAYER_DOMAIN] = controller
+    try:
+        await controller.connect(auto_reconnect=True)
 
-    hass.async_create_task(async_load_platform(
-        hass, MEDIA_PLAYER_DOMAIN, DOMAIN, {}, config))
+        async def disconnect_controller(event):
+            await controller.disconnect()
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP,
+                                   disconnect_controller)
+        hass.data[DOMAIN] = {
+            DATA_CONTROLLER: controller,
+            MEDIA_PLAYER_DOMAIN: await controller.get_players()
+        }
+    # Auto reconnect only operates if initial connection was successful.
+    except (asyncio.TimeoutError, ConnectionError):
+        await controller.disconnect()
+        _LOGGER.exception("Error connecting and retrieving players")
+        return False
 
+    hass.async_create_task(hass.config_entries.async_forward_entry_setup(
+        entry, MEDIA_PLAYER_DOMAIN))
     return True
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry):
```

Comment:
This isn't necessary for config entry support, lets keep this tight

Reply:
Huh?  Yes it is.

Reply:
Nu uh! ;) Fall back is that hass requires a restart to unload.

Reply:
And leave the HEOS to connection open meanwhile, which is unacceptable.  This should stay in, not be pushed to a new PR (it's 7 lines of code + 1 test).